### PR TITLE
fix: 'unexpected reserved bits' breaking web terminal (#9605)

### LINKED ISSUE #9605: OCI runtime exec failed: exec failed: unable to start container process: exec: "bash": executable file not found in $PATH: unknown
Terminal exec(v2.4.0-rc5) only support the docker image which entrypoint is "bash", like nginx. Could "sh" entrypoint (is widely used) be supported by stable version in the future?  THX!

**Terminal error messages:** 
OCI runtime exec failed: exec failed: unable to start container process: exec: "bash": executable file not found in $PATH: unknown

**F12:**
WebSocket is already in CLOSING or CLOSED state

### LINKED ISSUE #9641: Problem with Terminal and linux shell different from bash
**Describe the bug**

Tried the new most wanted feature, Terminal, it works with linux container with `bash` but got some problem with container `Alpine` based.

In screenshot, tried on `busybox:stable` container (it's in loop to keep it running) got # after `OCI runtime exec failed: exec failed: container_linux.go:380: starting container process caused: exec: "bash": executable file not found in $PATH: unknown` but cannot use it (cannot write)

Tried with some other container based on `node:lts-alpine` and got only black screen with blinking cursor.

Shell is working with same container using `Rancher` and `Lens`.

**To Reproduce**

Click on Terminal on Alpine based container

**Expected behavior**

Get shell from container and possibility to use it

**Screenshots**
image

**Version**

```shell
2.4.0+91aefab
```

### LINKED ISSUE #9643: ArgoCD v.2.4.0 Terminal read message err: websocket: unexpected reserved bits 0x70
My argocd-cm:
`apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: argocd-cm
    app.kubernetes.io/part-of: argocd
  name: argocd-cm
data:
  exec.enabled: 'true'`

My argocd-params-cm:
`apiVersion: v1
kind: ConfigMap
metadata:
  labels:
    app.kubernetes.io/name: argocd-cmd-params-cm
    app.kubernetes.io/part-of: argocd
  name: argocd-cmd-params-cm
data:
  server.insecure: 'true'
  server.enable.exec: 'true'`

My argocd-server cluster role:
`apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
    app.kubernetes.io/component: server
    app.kubernetes.io/name: argocd-server
    app.kubernetes.io/part-of: argocd
    argocd.argoproj.io/instance: argocd-itself
  name: argocd-server
rules:
- apiGroups:
  - '*'
  resources:
  - '*'
  verbs:
  - delete
  - get
  - patch
- apiGroups:
  - ""
  resources:
  - events
  verbs:
  - list
- apiGroups:
  - ""
  resources:
  - pods
  - pods/log
  verbs:
  - get
- apiGroups:
  - ""
  resources:
  - pods/exec
  verbs:
  - create`

I've enabled websockets on my HTTPProxy and I can see the terminal for a pod, but I can't type in the terminal. I see the error in the argocd-server:

time="2022-06-13T14:22:05Z" level=error msg="read message err: websocket: unexpected reserved bits 0x70"
E0613 14:22:05.001903       1 v2.go:105] websocket: unexpected reserved bits 0x70

### PR DESCRIPTION
**Pull request by crenshaw-dev**
Fixes #9605 
Fixes #9641
Fixes #9643

Hi I'm Michael, and I suck at concurrent programming.

**But** [another issue](https://github.com/gorilla/websocket/issues/373) directed me to run web terminal with the race detector enabled.

And I got a bunch of text that I didn't understand.

<details>
<summary>Very boring race detector output</summary>

```
==================
WARNING: DATA RACE
Write at 0x00c0016f0008 by goroutine 44:
  runtime.racewriterange()
      <autogenerated>:1 +0x29
  internal/poll.ignoringEINTRIO()
      /usr/local/go/src/internal/poll/fd_unix.go:794 +0x44b
  internal/poll.(*FD).Read()
      /usr/local/go/src/internal/poll/fd_unix.go:163 +0x26
  net.(*netFD).Read()
      /usr/local/go/src/net/fd_posix.go:55 +0x50
  net.(*conn).Read()
      /usr/local/go/src/net/net.go:183 +0xb0
  net.(*TCPConn).Read()
      <autogenerated>:1 +0x64
  github.com/soheilhy/cmux.(*bufferedReader).Read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/soheilhy/cmux/buffer.go:53 +0x2f7
  github.com/soheilhy/cmux.(*MuxConn).Read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/soheilhy/cmux/cmux.go:297 +0x50
  bufio.(*Reader).fill()
      /usr/local/go/src/bufio/bufio.go:106 +0x293
  bufio.(*Reader).Peek()
      /usr/local/go/src/bufio/bufio.go:144 +0xcb
  github.com/gorilla/websocket.(*Conn).read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/gorilla/websocket/conn.go:370 +0x5a
  github.com/gorilla/websocket.(*Conn).advanceFrame()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/gorilla/websocket/conn.go:798 +0x124
  github.com/gorilla/websocket.(*Conn).NextReader()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/gorilla/websocket/conn.go:980 +0x164
  github.com/gorilla/websocket.(*Conn).ReadMessage()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/gorilla/websocket/conn.go:1064 +0x30
  github.com/argoproj/argo-cd/v2/server/application.(*terminalSession).Read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/application/websocket.go:63 +0x69
  k8s.io/client-go/tools/remotecommand.readerWrapper.Read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/reader.go:40 +0x79
  k8s.io/client-go/tools/remotecommand.(*readerWrapper).Read()
      <autogenerated>:1 +0x29
  io.copyBuffer()
      /usr/local/go/src/io/io.go:426 +0x28a
  io.Copy()
      /usr/local/go/src/io/io.go:385 +0x1ab
  k8s.io/client-go/tools/remotecommand.(*streamProtocolV2).copyStdin.func1()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/v2.go:104 +0xfd

Previous read at 0x00c0016f0008 by goroutine 125:
  runtime.slicecopy()
      /usr/local/go/src/runtime/slice.go:295 +0x0
  bufio.(*Reader).Read()
      /usr/local/go/src/bufio/bufio.go:249 +0x6c4
  github.com/gorilla/websocket.(*messageReader).Read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/gorilla/websocket/conn.go:1021 +0x3cc
  io.ReadAll()
      /usr/local/go/src/io/io.go:645 +0x102
  io/ioutil.ReadAll()
      /usr/local/go/src/io/ioutil/ioutil.go:27 +0x75
  github.com/gorilla/websocket.(*Conn).ReadMessage()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/github.com/gorilla/websocket/conn.go:1068 +0x6a
  github.com/argoproj/argo-cd/v2/server/application.(*terminalSession).Read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/application/websocket.go:63 +0x69
  k8s.io/client-go/tools/remotecommand.readerWrapper.Read()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/reader.go:40 +0x79
  k8s.io/client-go/tools/remotecommand.(*readerWrapper).Read()
      <autogenerated>:1 +0x29
  io.copyBuffer()
      /usr/local/go/src/io/io.go:426 +0x28a
  io.Copy()
      /usr/local/go/src/io/io.go:385 +0x1ab
  k8s.io/client-go/tools/remotecommand.(*streamProtocolV2).copyStdin.func1()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/v2.go:104 +0xfd

Goroutine 44 (running) created at:
  k8s.io/client-go/tools/remotecommand.(*streamProtocolV2).copyStdin()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/v2.go:96 +0x11a
  k8s.io/client-go/tools/remotecommand.(*streamProtocolV4).stream()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/v4.go:65 +0x131
  k8s.io/client-go/tools/remotecommand.(*streamExecutor).Stream()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/remotecommand.go:141 +0xbd5
  github.com/argoproj/argo-cd/v2/server/application.startProcess()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/application/terminal.go:296 +0x4cc
  github.com/argoproj/argo-cd/v2/server/application.(*terminalHandler).ServeHTTP()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/application/terminal.go:232 +0x1fee
  github.com/argoproj/argo-cd/v2/server.(*ArgoCDServer).newHTTPServer.func1()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/server.go:836 +0x6fb
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  net/http.(*ServeMux).ServeHTTP()
      /usr/local/go/src/net/http/server.go:2462 +0xc5
  github.com/argoproj/argo-cd/v2/server.(*handlerSwitcher).ServeHTTP()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/server.go:1132 +0x1bc
  github.com/argoproj/argo-cd/v2/server.(*bug21955Workaround).ServeHTTP()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/server.go:1157 +0x235
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /usr/local/go/src/net/http/server.go:3071 +0x58

Goroutine 125 (finished) created at:
  k8s.io/client-go/tools/remotecommand.(*streamProtocolV2).copyStdin()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/v2.go:96 +0x11a
  k8s.io/client-go/tools/remotecommand.(*streamProtocolV4).stream()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/v4.go:65 +0x131
  k8s.io/client-go/tools/remotecommand.(*streamExecutor).Stream()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/vendor/k8s.io/client-go/tools/remotecommand/remotecommand.go:141 +0xbd5
  github.com/argoproj/argo-cd/v2/server/application.startProcess()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/application/terminal.go:296 +0x4cc
  github.com/argoproj/argo-cd/v2/server/application.(*terminalHandler).ServeHTTP()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/application/terminal.go:232 +0x1fee
  github.com/argoproj/argo-cd/v2/server.(*ArgoCDServer).newHTTPServer.func1()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/server.go:836 +0x6fb
  net/http.HandlerFunc.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2084 +0x4d
  net/http.(*ServeMux).ServeHTTP()
      /usr/local/go/src/net/http/server.go:2462 +0xc5
  github.com/argoproj/argo-cd/v2/server.(*handlerSwitcher).ServeHTTP()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/server.go:1132 +0x1bc
  github.com/argoproj/argo-cd/v2/server.(*bug21955Workaround).ServeHTTP()
      /Users/mcrenshaw/go/src/github.com/argoproj/argo-cd/server/server.go:1157 +0x235
  net/http.serverHandler.ServeHTTP()
      /usr/local/go/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /usr/local/go/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /usr/local/go/src/net/http/server.go:3071 +0x58
==================
```
</details>

Aimlessly clicking around led me to [this issue](https://github.com/gorilla/websocket/issues/400#issuecomment-408699663) linking to [some websocket documentation](https://pkg.go.dev/github.com/gorilla/websocket?utm_source=godoc#hdr-Concurrency) which seems to say that when we use `ReadMessage`, we gotta protect it with a mutex.

So I added a mutex around `ReadMessage` - and everything works! Conclusion: I am a genius who is excellent at concurrent programming.

This PR also makes the shell list/order configurable, because we need that anyway. I didn't change the order.

I've tested `debian:latest` and `alpine:latest`. Both work with the default shell order.

Should we also protect the `WriteMessage` call with a mutex? I'm not sure. I guess maybe wait until there's a bug to fix?

### CODE REVIEW

**Review by notfromstatefarm — CHANGES_REQUESTED** [review]
HALLELUJAH!

File: util/settings/settings.go [hunk 1]
```diff
@@ -1273,6 +1277,13 @@ func updateSettingsFromConfigMap(settings *ArgoCDSettings, argoCDCM *apiv1.Confi
 	}
 	settings.InClusterEnabled = argoCDCM.Data[inClusterEnabledKey] != "false"
 	settings.ExecEnabled = argoCDCM.Data[execEnabledKey] == "true"
+	execShells := argoCDCM.Data[execShellsKey]
+	if execShells != "" {
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply util/settings/settings.go:1281]: `Data` is `map[string]string, so if it's missing, it should just be `""`.

But I did realize I'm unnecessarily passing around a pointer to this array. I thought it was gonna auto-update the value when I changed the ConfigMap, but that didn't work. So I'm not gonna bother with that.

File: server/application/terminal.go [hunk 2]
```diff
@@ -123,7 +125,7 @@ func (s *terminalHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		http.Error(w, "Namespace name is not valid", http.StatusBadRequest)
 		return
 	}
-	shell := q.Get("shell") // No need to validate. Will only buse used if it's in the allow-list.
+	shell := q.Get("shell") // No need to validate. Will only use used if it's in the allow-list.
```
[the remark this answers is not in the document]
Comment (crenshaw-dev) [reply server/application/terminal.go:128]: lol oops

**Review by notfromstatefarm — APPROVED** [review]
LGTM! :shipit: 

I tested it in a live cluster and it works!

**Comment by jannfis** [comment]
> @yeya24 do you have time to take a glance at this? Any thoughts on also putting `WriteMessage` in a mutex?

According to https://pkg.go.dev/github.com/gorilla/websocket#hdr-Concurrency, reading and writing do not support concurrent access. So ensuring access is locked by a mutex seems generally a good idea.

However, something feels odd. Can the web terminal connections be muxed between clients? I was assuming that each client will have a unique socket for terminal access, and therefore, the socket shouldn't have concurrent access. I have not dived into the code tho.

**Comment by crenshaw-dev** [comment]
The way I'm reading this, net/http starts a goroutine for each connection received. I would _expect_ that there would be one connection per terminal session. However, I guess it's possible that there is one connection per message sent to the websocket. In that case, you'd have to goroutines pulling off the shared read buffer.

**Review by jannfis — APPROVED** [review]
LGTM with just a couple of nits. WDYT?

**Comment by crenshaw-dev** [comment]
Nitpicks applied! I'll plan to release this next Tuesday afternoon.

**Review by yeya24 — APPROVED** [review]
LGTM. Thanks for fixing this!

**Comment by michalziobro** [comment]
I'm still having this issue with 2.4.4 version:
argocd-server log:
```
E0712 10:17:06.186996       1 v3.go:79] EOF
9
E0712 10:17:06.429485       1 v2.go:105] EOF
8
time="2022-07-12T10:17:07Z" level=error msg="read message err: websocket: unexpected reserved bits 0x40"
7
E0712 10:17:07.727456       1 v2.go:105] websocket: unexpected reserved bits 0x40
```

blinking cursor in terminal tab, cannot type anything:
image

Terminal works ok with bash.

**Comment by crenshaw-dev** [comment]
Cherry-picked onto release-2.4 for 2.4.6.

**Pull request merged**